drivers/timer/sam0_rtc_timer.c

```diff
@@ -0,0 +1,290 @@
+#define RTC0 ((RtcMode0 *) DT_RTC_SAM0_BASE_ADDRESS)
+
+/* Number of sys timer cycles per on tick. */
+#define CYCLES_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC \
+			 / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+/* Maximum number of ticks. */
+#define MAX_TICKS (UINT32_MAX / CYCLES_PER_TICK - 2)
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/*
+ * Due to the nature of clock synchronization, reading from or writing to some
+ * RTC registers takes approximately six RTC_GCLK cycles. This constant defines
+ * a safe threshold for the comparator.
+ */
+#define TICK_THRESHOLD 7
+
+#else /* !CONFIG_TICKLESS_KERNEL */
+
+/*
+ * For some reason, RTC does not generate interrupts when COMP == 0,
+ * MATCHCLR == 1 and PRESCALER == 0. So we need to check that CYCLES_PER_TICK
+ * is more than one.
+ */
+#if CYCLES_PER_TICK < 2
+#error unsupported configuration for ticking mode; CYCLES_PER_TICK must be \
+	greater than 2
+#endif
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+/* Helper macro to get the correct GCLK GEN based on configuration. */
+#define GCLK_GEN(n) GCLK_EVAL(n)
+#define GCLK_EVAL(n) GCLK_CLKCTRL_GEN_GCLK##n
+
+/* Tick count of the last announce call. */
+static volatile u32_t rtc_last;
+
+#ifndef CONFIG_TICKLESS_KERNEL
+
+/* Current tick count. */
+static volatile u32_t rtc_counter;
+
+/* Tick value of the next timeout. */
+static volatile u32_t rtc_timeout;
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+static void rtc_reset(void);
+static void rtc_isr(void *arg);
+
+/*
+ * Waits for RTC bus synchronization.
+ */
+static inline void rtc_sync(void)
+{
+	while (RTC0->STATUS.reg & RTC_STATUS_SYNCBUSY) {
+		/* Wait for bus synchronization... */
+	}
+}
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/*
+ * Reads RTC COUNT register. First a read request must be written to READREQ,
+ * then - when bus synchronization completes - the COUNT register is read and
+ * returned.
+ */
+static u32_t rtc_count(void)
+{
+	/* Request synchronization and read the COUNT register. */
+	RTC0->READREQ.reg = RTC_READREQ_RREQ;
+	rtc_sync();
+	return RTC0->COUNT.reg;
+}
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+
+/*
+ * Initializes the RTC peripheral.
+ */
+int z_clock_driver_init(struct device *device)
+{
+	ARG_UNUSED(device);
+
+	/* Set up bus clock and GCLK generator. */
+	PM->APBAMASK.reg |= PM_APBAMASK_RTC;
+	GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(RTC_GCLK_ID) | GCLK_CLKCTRL_CLKEN
+			    | GCLK_GEN(DT_RTC_SAM0_CLOCK_GENERATOR);
+
+	while (GCLK->STATUS.bit.SYNCBUSY) {
+		/* Synchronize GCLK. */
+	}
+
+	/* Reset module to hardware defaults. */
+	rtc_reset();
+
+	rtc_last = 0;
+
+	/* Configure RTC with 32-bit mode, configured prescaler and MATCHCLR. */
+	u16_t ctrl = RTC_MODE0_CTRL_MODE(0) | RTC_MODE0_CTRL_PRESCALER(0);
+#ifndef CONFIG_TICKLESS_KERNEL
+	ctrl |= RTC_MODE0_CTRL_MATCHCLR;
+#endif
+	rtc_sync();
+	RTC0->CTRL.reg = ctrl;
+
+	/* Enable RTC interrupt. */
+	NVIC_ClearPendingIRQ(DT_RTC_SAM0_IRQ);
+	IRQ_CONNECT(DT_RTC_SAM0_IRQ, DT_RTC_SAM0_IRQ_PRIORITY, rtc_isr, 0, 0);
+	irq_enable(DT_RTC_SAM0_IRQ);
```

Comment:
You probably want to enable the IRQ last unless you have a really firm guarantee that the hardware initialization or firmware (or previous OS insance, maybe) didn't leave it in a state where the interrupt will fire before you get to the end of initialization.

Reply:
The `rtc_reset()` function performs software reset of the RTC peripheral, but... better safe than sorry I guess :-) I moved IRQ initialization to the end of the function.